File: matiasmasso/Mat.NET/Web/Web/Models/BoxViewModel.vb
Public Class BoxViewModel
    Property Title As String
    Property Text As String
    Property NavigateTo As String
    Property Tag As String

    Shared Function Factory(title As String, Optional navigateTo As String = "", Optional text As String = "", Optional tag As String = "") As BoxViewModel
        Dim retval As New BoxViewModel
        With retval
            .Title = title
            .NavigateTo = navigateTo
            .Text = text
            .Tag = tag
        End With
        Return retval
    End Function

    Public Class Collection
        Inherits List(Of BoxViewModel)

        Public Overloads Function Add(title As String, Optional navigateTo As String = "", Optional text As String = "", Optional tag As String = "") As BoxViewModel
            Dim retval = BoxViewModel.Factory(title, navigateTo, text, tag)
            MyBase.Add(retval)
            Return retval
        End Function
    End Class

End Class

Public Class ImageBoxViewModel
    Inherits BoxViewModel

    Property ImageUrl As String
    Property ImageWidth As Integer
    Property ImageHeight As Integer

    Overloads Shared Function Factory(imageUrl As String, imageWidth As Integer, imageHeight As Integer, title As String, Optional navigateTo As String = "", Optional text As String = "", Optional tag As String = "") As BoxViewModel
        Dim retval As New ImageBoxViewModel
        With retval
            .ImageUrl = imageUrl
            .ImageWidth = imageWidth
            .ImageHeight = imageHeight
            .Title = title
            .NavigateTo = navigateTo
            .Text = text
            .Tag = tag
        End With
        Return retval
    End Function

    Public Shadows Class Collection
        Inherits List(Of ImageBoxViewModel)

        Shadows Function Add(imageUrl As String, imageWidth As Integer, imageHeight As Integer, title As String, Optional navigateTo As String = "", Optional text As String = "", Optional tag As String = "") As BoxViewModel
            Dim retval = ImageBoxViewModel.Factory(imageUrl, imageWidth, imageHeight, title, navigateTo, text, tag)
            MyBase.Add(retval)
            Return retval
        End Function

        Shadows Function Add(title As String, Optional navigateTo As String = "", Optional text As String = "", Optional tag As String = "") As BoxViewModel
            Dim retval = ImageBoxViewModel.Factory(title, navigateTo, text, tag)
            MyBase.Add(retval)
            Return retval
        End Function

    End Class


End Class

File: matiasmasso/Mat.NET/DTO/Models/BoxViewModel.cs
using System.Collections.Generic;

namespace DTO
{
    public class BoxViewModel
    {
        public string Title { get; set; }
        public string Text { get; set; }
        public string NavigateTo { get; set; }
        public string Tag { get; set; }

        public static BoxViewModel Factory(string title, string navigateTo = "", string text = "", string tag = "")
        {
            BoxViewModel retval = new BoxViewModel();
            {
                var withBlock = retval;
                withBlock.Title = title;
                withBlock.NavigateTo = navigateTo;
                withBlock.Text = text;
                withBlock.Tag = tag;
            }
            return retval;
        }

        public class Collection : List<BoxViewModel>
        {
            public BoxViewModel Add(string title, string navigateTo = "", string text = "", string tag = "")
            {
                var retval = BoxViewModel.Factory(title, navigateTo, text, tag);
                base.Add(retval);
                return retval;
            }
        }

        public override string ToString()
        {
            return string.Format("BoxViewModel: {0} -> {1}", Title ?? "", NavigateTo ?? "");
        }
    }

    public class ImageBoxViewModel : BoxViewModel
    {
        public string ImageUrl { get; set; }
        public int ImageWidth { get; set; }
        public int ImageHeight { get; set; }


        public static new ImageBoxViewModel Factory(string title, string navigateTo = "", string text = "", string tag = "")
        {
            ImageBoxViewModel retval = new ImageBoxViewModel();
            {
                var withBlock = retval;
                withBlock.Title = title;
                withBlock.NavigateTo = navigateTo;
                withBlock.Text = text;
                withBlock.Tag = tag;
            }
            return retval;
        }


        public static ImageBoxViewModel Factory(string imageUrl, int imageWidth, int imageHeight, string title, string navigateTo = "", string text = "", string tag = "")
        {
            ImageBoxViewModel retval = new ImageBoxViewModel();
            {
                var withBlock = retval;
                withBlock.ImageUrl = imageUrl;
                withBlock.ImageWidth = imageWidth;
                withBlock.ImageHeight = imageHeight;
                withBlock.Title = title;
                withBlock.NavigateTo = navigateTo;
                withBlock.Text = text;
                withBlock.Tag = tag;
            }
            return retval;
        }

        public new class Collection : List<ImageBoxViewModel>
        {
            public ImageBoxViewModel Add(string imageUrl, int imageWidth, int imageHeight, string title, string navigateTo = "", string text = "", string tag = "")
            {
                ImageBoxViewModel retval = ImageBoxViewModel.Factory(imageUrl, imageWidth, imageHeight, title, navigateTo, text, tag);
                base.Add(retval);
                return retval;
            }

            public ImageBoxViewModel Add(string title, string navigateTo = "", string text = "", string tag = "")
            {
                ImageBoxViewModel retval = ImageBoxViewModel.Factory(title, navigateTo, text, tag);
                base.Add(retval);
                return retval;
            }
        }
    }
}
